homeassistant/components/tplink_omada/update.py

```diff
@@ -0,0 +1,149 @@
+    omada_client = controller.omada_client
+
+    devices = await omada_client.get_devices()
+
+    coordinator = OmadaCoordinator[FirmwareUpdateStatus](
+        hass,
+        omada_client,
+        "Firmware Updates",
+        _poll_firmware_updates,
+        poll_delay=6 * 60 * 60,
+    )
+
+    entities: list = []
+    for device in devices:
+        entities.append(OmadaDeviceUpdate(coordinator, device))
+
+    async_add_entities(entities)
+    await coordinator.async_request_refresh()
+
+
+class OmadaDeviceUpdate(
+    OmadaDeviceEntity[FirmwareUpdateStatus],
+    UpdateEntity,
+):
+    """Firmware update status for Omada SDN devices."""
+
+    _attr_supported_features = (
+        UpdateEntityFeature.INSTALL
+        | UpdateEntityFeature.PROGRESS
+        | UpdateEntityFeature.RELEASE_NOTES
+    )
+    _firmware_update: OmadaFirmwareUpdate = None
+
+    def __init__(
+        self,
+        coordinator: OmadaCoordinator[FirmwareUpdateStatus],
+        device: OmadaListDevice,
+    ) -> None:
+        """Initialize the update entity."""
+        super().__init__(coordinator, device)
+
+        self._mac = device.mac
+        self._device = device
+        self._omada_client = coordinator.omada_client
+
+        self._attr_unique_id = f"{device.mac}_firmware"
+        self._attr_has_entity_name = True
+        self._attr_name = "Firmware Update"
+        self._refresh_state()
+
+    def _refresh_state(self) -> None:
+        if self._firmware_update and self._device.need_upgrade:
+            self._attr_installed_version = self._firmware_update.current_version
+            self._attr_latest_version = self._firmware_update.latest_version
+        else:
+            self._attr_installed_version = self._device.firmware_version
+            self._attr_latest_version = self._device.firmware_version
+        self._attr_in_progress = self._device.fw_download
+
+        if self._attr_in_progress:
+            # While firmware update is in progress, poll more frequently
+            async_call_later(self.hass, 60, self._request_refresh)
+
+    async def _request_refresh(self, _now: Any) -> None:
+        await self.coordinator.async_request_refresh()
+
+    def release_notes(self) -> str | None:
+        """Get the release notes for the latest update."""
+        if self._firmware_update:
+            return str(self._firmware_update.release_notes)
+        return ""
+
+    async def async_install(
+        self, version: str | None, backup: bool, **kwargs: Any
+    ) -> None:
+        """Install a firmware update."""
+        if self._firmware_update and (
+            version is None or self._firmware_update.latest_version == version
+        ):
+            await self._omada_client.start_firmware_upgrade(self._device)
+            await self.coordinator.async_request_refresh()
+        else:
+            _LOGGER.error("Firmware upgrade is not available for %s", self._device.name)
```

Comment:
This line seems unreachable to me.
As this is already cought by Home Assistant?